services/repository/branch.go

```diff
@@ -196,3 +360,66 @@ func DeleteBranch(ctx context.Context, doer *user_model.User, repo *repo_model.R
 
 	return nil
 }
+
+type BranchSyncOptions struct {
+	RepoID int64
+	DoerID int64
+}
+
+// branchSyncQueue represents a queue to handle branch s
+var branchSyncQueue *queue.WorkerPoolQueue[*BranchSyncOptions]
+
+// handle passed PR IDs and test the PRs
+func handlerBranchSync(items ...*BranchSyncOptions) []*BranchSyncOptions {
+	var failedOptions []*BranchSyncOptions
+	for _, opts := range items {
+		if err := repo_module.SyncRepoBranches(graceful.GetManager().ShutdownContext(), opts.RepoID, opts.DoerID); err != nil {
+			log.Error("syncRepoBranches [%d:%d] failed: %v", opts.RepoID, opts.DoerID, err)
+			failedOptions = append(failedOptions, opts)
+		}
+	}
+	return failedOptions
+}
+
+func addRepoToBranchSyncQueue(repoID, doerID int64) error {
+	return branchSyncQueue.Push(&BranchSyncOptions{
+		RepoID: repoID,
+		DoerID: doerID,
+	})
+}
+
+func initBranchSyncQueue(ctx context.Context) error {
+	branchSyncQueue = queue.CreateSimpleQueue(ctx, "branch_sync", handlerBranchSync)
+	if branchSyncQueue == nil {
+		return errors.New("unable to create branch_sync queue")
+	}
+	go graceful.GetManager().RunWithCancel(branchSyncQueue)
+
+	// for test env, this will return empty because data hasn't been loaded yet
+	// but that env don't need the sync
+	cnt, err := git_model.CountBranches(ctx, git_model.FindBranchOptions{
+		IsDeletedBranch: util.OptionalBoolFalse,
+	})
+	if err != nil {
+		return fmt.Errorf("CountBranches: %v", err)
+	}
+
+	if cnt == 0 {
+		go func() {
+			// doerID is 0 means no pusher will be displayed in the branch list ui
+			if err := AddAllRepoBranchesToSyncQueue(ctx, 0); err != nil {
+				log.Error("AddAllRepoBranchesToSyncQueue: %v", err)
+			}
+		}()
+	}
```

Comment:
It seems quite hacky. Without this block, what's the bad case?

Reply:
Hm, then the sync will be happened when the first user visit the repository home page.

Reply:
I think the sync already happen in `RepoAssignment`, that's good enough? Then no need this "AddAllRepoBranchesToSyncQueue"

TBH, without a real bad case, I do not think it's worth to make the system so complex, especially no test code for it.

Reply:
For those big repositories, the first-time the sync may take a lot of time. And this operation is equal to the administrator click the sync branch button on the admin panel.

Reply:
I do not think such big repo would take a lot of time.

Because, without this PR, these repositories already loads the branches again and again for each view. I think this PR won't make the first-time visit noticeable slower then before.

Reply:
Yes, so this PR will make it better in some cases. If the backend sync is finished before the user can visit the repository home page.

Maybe we can just sync according repository size.

Reply:
I really dislike the untested hacky code, and I don't see how it could be "better", users won't feel it.

Reply:
It will make the first time visit to a repository with over thousands of branches is still very fast.

Reply:
"still very fast": how slow was before and how much faster is now?

Even without this code, it's still not worse than before, and it's just for one-time use, so no need to introduce it?

Reply:
Let make an assumption: 

* git lists 1000 branches: 1 second
* insert 1000 records into database: 1 second

I do not think saving these 2 seconds for first-time visit is noticeable to end users, a uncontrollable network lag or disk io heavy load would cause longer time than it. Especially it's only a one-time operation, does any one really need it?

But introducing a queue is not cheap, especially the queue system is still not well-designed at the moment.

Reply:
All of them removed.